app/src/main/java/org/thoughtcrime/securesms/groups/GroupsV1MigrationUtil.java

```
@@ -0,0 +1,244 @@
+      Log.i(TAG, "Sending no-op update to notify others.");
+      GroupManager.sendNoopUpdate(context, gv2MasterKey, decryptedGroup);
+    }
+  }
+
+  public static void performLocalMigration(@NonNull Context context, @NonNull GroupId.V1 gv1Id) throws IOException
+  {
+    Log.i(TAG, "Beginning local migration! V1 ID: " + gv1Id, new Throwable());
+    try (Closeable ignored = GroupsV2ProcessingLock.acquireGroupProcessingLock()) {
+      if (DatabaseFactory.getGroupDatabase(context).groupExists(gv1Id.deriveV2MigrationGroupId())) {
+        Log.w(TAG, "Group was already migrated! Could have been waiting for the lock.", new Throwable());
+        return;
+      }
+
+      Recipient recipient = Recipient.externalGroupExact(context, gv1Id);
+      long      threadId  = DatabaseFactory.getThreadDatabase(context).getOrCreateThreadIdFor(recipient);
+
+      performLocalMigration(context, gv1Id, threadId, recipient);
+      Log.i(TAG, "Migration complete! (" + gv1Id + ", " + threadId + ", " + recipient.getId() + ")", new Throwable());
+    } catch (GroupChangeBusyException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private static @Nullable DecryptedGroup performLocalMigration(@NonNull Context context,
+                                                                @NonNull GroupId.V1 gv1Id,
+                                                                long threadId,
+                                                                @NonNull Recipient groupRecipient)
+      throws IOException, GroupChangeBusyException
+  {
+    Log.i(TAG, "performLocalMigration(" + gv1Id + ", " + threadId + ", " + groupRecipient.getId());
+
+    try (Closeable ignored = GroupsV2ProcessingLock.acquireGroupProcessingLock()){
+      DecryptedGroup decryptedGroup;
+      try {
+        decryptedGroup = GroupManager.addedGroupVersion(context, gv1Id.deriveV2MigrationMasterKey());
+      } catch (GroupDoesNotExistException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)